src/TableFunctions/ITableFunction.h

```diff
@@ -82,7 +83,7 @@ class ITableFunction : public std::enable_shared_from_this<ITableFunction>
 
     /// Create storage according to the query.
     StoragePtr
-    execute(const ASTPtr & ast_function, ContextPtr context, const std::string & table_name, ColumnsDescription cached_columns_ = {}, bool use_global_context = false, bool is_insert_query = false) const;
+    execute(const ASTPtr & ast_function, ContextPtr context, const std::string & table_name, ColumnsDescription cached_columns_ = {}, bool use_global_context = false, ASTInsertQuery * insert_query = nullptr) const;
```

Comment:
This API change is necessary because I want to be able to grab the partition by expression during the creation of `StorageObjectStorage` and not only during the call to `write`